File: cses/additional_problems/cyclic_array.py
n, k = map(int, input().split())
a = list(map(int, input().split()))

K = 19
p = -1
sum_val = 0
F = [[0] * (2 * n + 5) for _ in range(K)]

for i in range(2 * n):
    while p < 2 * n and sum_val <= k:
        sum_val += a[p + 1]
        p += 1
    F[0][i] = p
    sum_val -= a[i]

F[0][2 * n] = 2 * n

for k in range(1, K):
    for i in range(2 * n + 1):
        F[k][i] = F[k - 1][F[k - 1][i]]

res = n

for i in range(n):
    resi, x = 1, i
    for k in range(K - 1, -1, -1):
        if F[k][x] < i + n:
            x = F[k][x]
            resi += 1 << k
    res = min(res, resi)

print(res)

File: cses/additional_problems/cyclic_array.cpp
#include <bits/stdc++.h>
#define pb push_back
#define fst first
#define snd second
#define fore(i,a,b) for(ll i=a,ggdem=b;i<ggdem;++i)
#define SZ(x) ((int)x.size())
#define ALL(x) x.begin(),x.end()
#define mset(a,v) memset((a),(v),sizeof(a))
#define FIN ios::sync_with_stdio(0);cin.tie(0);cout.tie(0)
#define imp(v) for(auto messi:v)cout<<messi<<" "; cout<<"\n"
using namespace std;
typedef long long ll;
typedef pair<ll,ll> ii;
const ll K=19;
int main(){FIN;
	ll n,k; cin>>n>>k;
	vector<ll>a(2*n+5);
	fore(i,0,n)cin>>a[i],a[i+n]=a[i];
	ll p=-1,sum=0;
	ll F[K][2*n+5];
	fore(i,0,2*n){
		while(p<2*n&&sum<=k)sum+=a[++p];
		F[0][i]=p;
		sum-=a[i];
	}
	F[0][2*n]=2*n;
	//fore(i,0,2*n)cout<<F[0][i]<<" ";;cout<<endl;
	fore(k,1,K)fore(i,0,2*n+1)F[k][i]=F[k-1][F[k-1][i]];
	ll res=n;
	fore(i,0,n){
		ll resi=1,x=i;
		for(ll k=K-1;k>=0;k--)if(F[k][x]<i+n)x=F[k][x],resi+=1ll<<k;
		res=min(res,resi);
	}
	cout<<res<<"\n";
	return 0;
}
